updated Koperator documentation (#107)
banzaicloud/koperator-docs: docs/scenarios.md

---
title: Koperator capablities
weight: 400
---

As highlighted in the [features section]({{< relref "../_index.md#features" >}}), {{< kafka-operator >}} removed the reliance on StatefulSet,and supports several different usecases.

> Note: This is not a complete list, if you have a specific requirement or question, see our [support]({{< relref "/sdm/koperator/support.md">}}) options.

## Vertical capacity scaling

> You may have encountered situations where the horizontal scaling of a cluster is impossible. When **only one Broker is throttling** and needs more CPU or requires additional disks (because it handles the most partitions), a StatefulSet-based solution is useless, since it does not distinguish between replicas' specifications. The handling of such a case requires *unique* Broker configurations. If there is a need to add a new disk to a unique Broker, there can be waste of disk space (and money) with a StatefulSet-based solution, since it can't add a disk to a specific Broker, the StatefulSet adds one to each replica.

Reviewer: **Optional**: We **MAY** consider the following phrasing.

```suggestion
You may have encountered situations where the horizontal scaling of a cluster is impossible. When **only one Broker is throttling** and needs more CPU or requires additional disks (because it handles the most partitions), a StatefulSet-based solution is useless, since it does not distinguish between replicas' specifications. The handling of such a case requires *unique* Broker configurations. If there is a need to add a new disk to a unique Broker, there can be a waste of disk space (and money) with a StatefulSet-based solution, since it can't add a disk to a specific Broker, the StatefulSet adds one to each replica.
```

To remove the #1 Broker from the cluster, you need to lower the number of brokers in the cluster from three to one. This will cause a state in which only one Broker is live, while you kill the brokers that handle traffic. {{< kafka-operator >}} supports removing specific brokers without disrupting traffic in the cluster.

## Fine grained Broker config support

Apache Kafka is a stateful application, where Brokers create/form a cluster with other Brokers. Every Broker is uniquely configurable ({{< kafka-operator >}} supports heterogenous environments, in which no nodes are the same, act the same or have the same specifications - from the infrastructure up through the Brokers' Envoy configuration). Kafka has lots of Broker configs, which can be used to fine tune specific brokers, and {{< kafka-operator >}} did not want to limit these to ALL Brokers in a StatefulSet. {{< kafka-operator >}} supports unique Broker configs.

> *In each of the three scenarios lister above, {{< kafka-operator >}} does not use StatefulSet, relying, instead, on Pods, PVCs and ConfigMaps. With {{< kafka-operator >}}, StatefulSet is very convenient starting point, as it handles roughly 80% of scenarios but introduces huge limitations when running Kafka on Kubernetes in production.*

Reviewer: **Required**: We **MUST** clarify the wording here, Koperator doesn't use statefulset, other solutions using statefulset are fine to cover 80% of the cases, Koperator aims to cover 100%.

```suggestion
*In each of the three scenarios lister above, {{< kafka-operator >}} does not use StatefulSet, relying, instead, on Pods, PVCs and ConfigMaps. While using StatefulSet is a very convenient starting point, as it handles roughly 80% of scenarios, it also introduces huge limitations when running Kafka on Kubernetes in production.*
```

## Monitoring based control

Use of monitoring is essential for any application, and all relevant information about Kafka should be published to a monitoring solution. When using Kubernetes, the de facto solution is Prometheus, which supports configuring alerts based on previously consumed metrics. {{< kafka-operator >}} was built as a standards-based solution (Prometheus and Alert Manager) that could handle and react to alerts automatically, so human operators wouldn't have to. {{< kafka-operator >}} supports alert-based Kafka cluster management.

## LinkedIn's Cruise Control

>   LinkedIn knows how to operate Kafka in a better way. They built a tool, called Cruise Control, to operate their Kafka infrastructure. And {{< kafka-operator >}} is built to **handle the infrastructure, but not to reinvent the wheel in so far as operating Kafka**. {{< kafka-operator >}} is an operator which leveraged deep Kubernetes expertise by handling all Kafka infrastructure related issues in the best possible way. Managing Kafka can be a separate issue, for which there already exist some unique tools and solutions that are standard across the industry, so LinkedIn's Cruise Control is integrated with the {{< kafka-operator >}}.

Reviewer: **Required**: We **MUST** remove the leading spaces, they are unnecessary.

```suggestion
LinkedIn knows how to operate Kafka in a better way. They built a tool, called Cruise Control, to operate their Kafka infrastructure. And {{< kafka-operator >}} is built to **handle the infrastructure, but not to reinvent the wheel in so far as operating Kafka**. {{< kafka-operator >}} is an operator which leveraged deep Kubernetes expertise by handling all Kafka infrastructure related issues in the best possible way. Managing Kafka can be a separate issue, for which there already exist some unique tools and solutions that are standard across the industry, so LinkedIn's Cruise Control is integrated with the {{< kafka-operator >}}.
```

>   LinkedIn knows how to operate Kafka in a better way. They built a tool, called Cruise Control, to operate their Kafka infrastructure. And {{< kafka-operator >}} is built to **handle the infrastructure, but not to reinvent the wheel in so far as operating Kafka**. {{< kafka-operator >}} is an operator which leveraged deep Kubernetes expertise by handling all Kafka infrastructure related issues in the best possible way. Managing Kafka can be a separate issue, for which there already exist some unique tools and solutions that are standard across the industry, so LinkedIn's Cruise Control is integrated with the {{< kafka-operator >}}.

Reviewer: **Recommended**: The first sentence sounds like we were inferior to CC while that's not the case, we are complementary tools to each other, we **SHOULD** phrase this accordingly.

```suggestion
  LinkedIn knows how to operate Kafka at scale. They built a tool, called Cruise Control, to operate their Kafka infrastructure. And {{< kafka-operator >}} is built to **handle the infrastructure, but not to reinvent the wheel in so far as operating Kafka**. {{< kafka-operator >}} is an operator which leveraged deep Kubernetes expertise by handling all Kafka infrastructure related issues in the best possible way. Managing Kafka can be a separate issue, for which there already exist some unique tools and solutions that are standard across the industry, so LinkedIn's Cruise Control is integrated with the {{< kafka-operator >}}.
```

>   LinkedIn knows how to operate Kafka in a better way. They built a tool, called Cruise Control, to operate their Kafka infrastructure. And {{< kafka-operator >}} is built to **handle the infrastructure, but not to reinvent the wheel in so far as operating Kafka**. {{< kafka-operator >}} is an operator which leveraged deep Kubernetes expertise by handling all Kafka infrastructure related issues in the best possible way. Managing Kafka can be a separate issue, for which there already exist some unique tools and solutions that are standard across the industry, so LinkedIn's Cruise Control is integrated with the {{< kafka-operator >}}.

Reviewer: **Optional**: We **MAY** want to improve phrasing.

```suggestion
  LinkedIn knows how to operate Kafka in a better way. They built a tool, called Cruise Control, to operate their Kafka infrastructure. And {{< kafka-operator >}} is built to **handle the infrastructure, but not to reinvent the wheel in so far as operating Kafka**. {{< kafka-operator >}} was built to leverage the Kubernetes operator pattern and our Kubernetes expertise by handling all Kafka infrastructure related issues in the best possible way. Managing Kafka can be a separate issue, for which there already exist some unique tools and solutions that are standard across the industry, so LinkedIn's Cruise Control is integrated with the {{< kafka-operator >}}.
```
